### PR TITLE
[sensors][android] Add Light sensor

### PR DESCRIPTION
**Pull request by bearkillerPT**
# Introduction

Hey there! You can check more about me on my profile but I'm kind of new to contribuiting to such an insanely huge and useful projects as yours! I hope I don't make many mistakes by trying to introduce this feature. Thank you for your help!
In this PR I created a clean for and a working branch for easy integration!
# Why

I'm working in a project with expo and wanted to have access to the Light Sensor for my app. 
Also, if you're interested, refer to [this PR](https://github.com/expo/expo/pull/18177)!
# How

I build this feature based on the discussion and guidance from @barthap [here](https://github.com/expo/expo/discussions/18101). I mostly copied the barometer's implementation.

# Test Plan

The test can be acomplished as the new doc entries describe and similarly to the barometer.
 [light-sensor.md](https://github.com/bearkillerPT/expo/blob/main/docs/pages/versions/unversioned/sdk/light-sensor.md).

For the standalone PR test you can use the code in those same docs and run yarn.
expo install expo-sensors.
expo eject.
Remove expo-modules-core and expo-sensors from node_modules.
Afterwards copy both the expo-modules-core and expo-sensors to node_modules.
cd android;
./gradlew assembleRelease //Generate the apk in 
app/build/outputs/apk/release;

Send it to your device and test it!

# Checklist

- [x ] Documentation is up to date to reflect these changes [light-sensor.md](https://github.com/bearkillerPT/expo/blob/main/docs/pages/versions/unversioned/sdk/light-sensor.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).
I'm not sure about this last two... But everything seems to work fine!

### CODE REVIEW

**Review by barthap — CHANGES_REQUESTED** [review]
Ok, two more things to do (in addition to my comments):
- Go into `packages/expo-sensors` directory and run `yarn build` again. 
- Go into `apps/bare-expo/android` directory and run `./gradlew :expo-sensors:spotlessApply` - we have linter for Kotlin files and they need to be reformatted

Then commit changes made by these tools

File: docs/pages/versions/v46.0.0/sdk/light-sensor.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+---
+title: LightSensor
+sourceCodeUrl: 'https://github.com/expo/expo/tree/main/packages/expo-sensors'
```
[the remark this answers is not in the document]
Comment (bearkillerPT) [reply docs/pages/versions/v46.0.0/sdk/light-sensor.md]: Okay done!

**Comment by bearkillerPT** [comment]
> Ok, two more things to do (in addition to my comments):
> 
> * Go into `packages/expo-sensors` directory and run `yarn build` again.
> * Go into `apps/bare-expo/android` directory and run `./gradlew :expo-sensors:spotlessApply` - we have linter for Kotlin files and they need to be reformatted
> 
> Then commit changes made by these tools

Alright! That's done and commited!

**Comment by bearkillerPT** [comment]
> Great! grin Only changelog left - you can mostly copy-paste the changes from the previous PR (remember to update the numbes in link). Please do change the file in `packages/expo-sensors/CHANGELOG.md` wink

Oops! I forgot about the CAHNGELOG! I've commited it with the new PR link!

**Comment by bearkillerPT** [comment]
> _Hi there! wave I'm a bot whose goal is to ensure your contributions meet our guidelines._
> 
> I've found some issues in your pull request that should be addressed (click on them for more details) point_down
> 
> **warning Suggestion**: Missing changelog entries
> Your changes should be noted in the changelog. Read [Updating Changelogs](https://github.com/expo/expo/blob/main/guides/contributing/Updating%20Changelogs.md) guide and consider adding an appropriate entry to the following changelogs:
> 
> * [`packages/expo-modules-core/CHANGELOG.md`](https://github.com/bearkillerPT/expo/blob/LightSensor/packages/expo-modules-core/CHANGELOG.md)
> 
> _Generated by ExpoBot robot against [c652be9](https://github.com/expo/expo/commit/c652be92e5ed72aaeb0ea9956677784337285662)_

I was thinking about adding:
## Unpublished

- Added the interface for the light sensor support on Android. ([#18225](https://github.com/expo/expo/pull/18225) by [bearkillerPT](https://github.com/bearkillerPT))
Should I?

**Comment by barthap** [comment]
I don't know either if we should add changelog for expo-modules-core too 😅 cc @tsapeta

**Comment by tsapeta** [comment]
@bearkillerPT, what you proposed seems fine and correct to me 😄
By the way, great work on this PR! 👏

**Comment by bearkillerPT** [comment]
> @bearkillerPT, what you proposed seems fine and correct to me smile By the way, great work on this PR! clap

Thank you! My pleasure :)

File: packages/expo-sensors/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 ## Unpublished
 
+- Added support for the light sensor on Android. ([#18225](https://github.com/expo/expo/pull/18225) by [bearkillerPT](https://github.com/bearkillerPT))
```
Comment (barthap) [inline packages/expo-sensors/CHANGELOG.md:5]: Now I noticed it,

Please move it under "New features" section 😉 

Spacing here is perfect - one blank line above and one below 🤩 keep it this way

File: packages/expo-modules-core/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 ## Unpublished
 
+
+- Added the interface for the light sensor support on Android. ([#18225](https://github.com/expo/expo/pull/18225) by [bearkillerPT](https://github.com/bearkillerPT))
```
Comment (barthap) [inline packages/expo-modules-core/CHANGELOG.md:6]: Could you move this to the "New features" section too? And also, keep the blank lines the same way as you did in `expo-sensors/CHANGELOG.md` - it looks perfect there

**Review by barthap — COMMENTED** [review]
I'm afraid we're going to ship it to Expo Go with SDK 47, because we're already started the sdk-46 release process of it, and this one would need a few more changes to Expo Go (out of scope of this PR).
But most likely, you'll be able to use it with a [custom development client](https://docs.expo.dev/development/introduction/) and EAS Build soon.

**Comment by bearkillerPT** [comment]
> I'm afraid we're going to ship it to Expo Go with SDK 47, because we're already started the sdk-46 release process of it, and this one would need a few more changes to Expo Go (out of scope of this PR). But most likely, you'll be able to use it with a [custom development client](https://docs.expo.dev/development/introduction/) soon.

So if I wish to use it with my project do I simply build the App by replacihng these packages? I read the link but I have an app on the Play Store...

**Review by barthap — APPROVED** [review]
Looks good to me 🎉 Great job @bearkillerPT 👏 

_Note to other reviewers: I'm not sure if we want to merge it before sdk-46 cutoff, in `expoview` there is some magic like `ScopedSensorService` which was not implemented in this PR, so I'm not sure if this sensor will be working in Expo Go without any further modifications. I'll test it and implement missing interfaces in a subsequent PR._

**Comment by barthap** [comment]
> So if I wish to use it with my project do I simply build the App by replacing these packages? I read the link but I have an app on the Play Store...

Briefly:
- If you're using classic builds (`expo build`) - then no. this won't be possible anymore, because SDK 47 stops supporting it
- If you're using EAS Build - this can be possible even today (I think you can patch `expo-sensors` and `expo-modules-core` with patch-package), but you won't be able to use it in Expo Go until SDK 47 - that's why I suggested that custom dev-client from that link 😉

**Comment by bearkillerPT** [comment]
> * patch-package

Hey! I've tried building the dev-client and then start --dev-client but i get the following error:
thumbnail_Screenshot_2022-07-21-11-01-46

Furthermore I tried to install bot patch-package and postinstall. Followed by deleting and replacing the expo-sensors and expo-modules-core (that i assume following this PR are already built) and try to run yarn patch-package expo-sensors:
image

Could you please help me? Being able to build and deploy is more important currently then being able to use de dev-client beacuse our app is almost finished. I would be happy to have both tho ;) 

PS: I commited recently to expo-sensors the index file as it wasn't exporting the light sensor yet!

File: docs/pages/versions/unversioned/sdk/light-sensor.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+packageName: 'expo-sensors'
+---
+
+import {APIInstallSection} from '~/components/plugins/InstallSection';
+import PlatformsSection from '~/components/plugins/PlatformsSection';
+import SnackInline from '~/components/plugins/SnackInline';
+
+import { InlineCode } from '~/components/base/code';
+
+`LightSensor` from **`expo-sensors`** provides access to the device light sensor to respond to changes in illuminance. `illuminance` is measured in _`Lux`_ or _`lx`_.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/light-sensor.md:13]:
```suggestion
`LightSensor` from **`expo-sensors`** provides access to the device's light sensor to respond to illuminance changes. `illuminance` is measured in _`Lux`_ or _`lx`_.
```

Comment (amandeepmittal) [reply docs/pages/versions/unversioned/sdk/light-sensor.md:13]: Just a small readability improvement 👍

**Review by amandeepmittal — COMMENTED** [review]
Left a suggestion for just a few small text improvements. 

Great work on this PR 👏

**Review by amandeepmittal — APPROVED** [review]

**Comment by bearkillerPT** [comment]
Hey! I need your help. I'm trying to reproduce my test of the last [pr](https://github.com/expo/expo/pull/18177) and i can't.
image
@barthap could you please take a look? I did the exact same steps and it says BUILD SUCCESFUL after expo run:android...

**Comment by barthap** [comment]
I'll have more time to look into this on Monday, but what I suspect is that your native changes aren't there when building the project. Or some of them are missing.

**Comment by bearkillerPT** [comment]
> I'll have more time to look into this on Monday, but what I suspect is that your native changes aren't there when building the project. Or some of them are missing.

Yeh I think some might be missing but I've checked and rechecked and can't find what it is... Haave you had the chance to look at it?

**Comment by bearkillerPT** [comment]
@barthap I’m extremely confused. I’ve already tried checking the commits of the last PR and still can’t get it to work. I’ve ran yarn in both expo-sensors and expo-modules-core and then copying it to the LighSensorTest project and running expo run:Android always gives me the same error: 

> Hey! I need your help. I'm trying to reproduce my test of the last [pr](https://github.com/expo/expo/pull/18177) and i can't.
> I've already tried the gh pr checkout and got the same result:
> image
> @barthap could you please take a look? I did the exact same steps and it says BUILD SUCCESFUL after expo run:android...

**Comment by barthap** [comment]
Hello, I apologize for not being responsible last week.

I checked out your PR and tested it in both Expo Go (built locally) and a bare-expo app on real Android device - it works perfectly 👏 
I'm going to merge it today - we're probably going to publish this package soon so you won't have to deal with patch-package etc. (of course in custom dev clients - Expo Go support will come with SDK 47)

This said, it must be something with your setup/configuration - it's hard to tell what exactly. If you have Android Studio, you can open `your_app/android` project directory, verify if your changes exist and try running your app from the IDE (then instead of `expo run:android` use `expo start --dev-client`).

**Comment by bearkillerPT** [comment]
> Hello, I apologize for not responding to your messages last week.
> 
> I checked out your PR and tested it in both Expo Go (built locally) and a bare-expo app on a real Android device - it works perfectly clap I'm going to merge it today - we're probably going to publish this package soon so you won't have to deal with patch-package etc. (of course in custom dev clients - Expo Go support will come with SDK 47)
> 
> This said, it must be something with your setup/configuration - it's hard to tell what exactly. If you have Android Studio, you can open `your_app/android` project directory, verify if your changes exist and try running your app from the IDE (then instead of `expo run:android` use `expo start --dev-client`).

image
Hey, @barthap, I'm sorry to still be bothering you with this but I've tried and tried and still can't get it to work. I cloned the expo repo went to packages/expo-sensors and packages/expo-modules-core and ran:
``` bash
yarn;
yarn build;
```
Afterwared I switched to my LightSensorTest and removed both from node_modules and copied the new ones. I've no idea of what is worng... do you think you could explain to me by steps how to do it?

Thank you for your time!

**Comment by bearkillerPT** [comment]
Hey! Thanks for the sdk 47 beta1! I've now tested with a development client and it works!

**Comment by barthap** [comment]
Yay! It should also work in Expo Go once the SDK 47 is fully released (most likely this week).

**Pull request merged**